lib/spack/spack/solver/concretize.lp

```diff
@@ -779,13 +779,18 @@ target_compatible(Descendent, Ancestor)
 #defined target_satisfies/2.
 #defined target_parent/2.
 
-% The target weight is either the default target weight
-% or a more specific per-package weight if set
+% If the package does not have any specific weight for this
+% target, offset the default weights by the number of specific
+% weights and use that. We additionally offset by 30 to ensure
+% preferences are propagated even against large numbers of
+% otherwise "better" matches.
 target_weight(Target, Package, Weight)
-  :- default_target_weight(Target, Weight),
+  :- default_target_weight(Target, DefaultWeight),
      node(Package),
      not derive_target_from_parent(_, Package),
-     not package_target_weight(Target, Package, _).
+     not package_target_weight(Target, Package, _),
+     Offset = #count{W : package_target_weight(_, Package, W)},
```

Comment:
I agree with the PR but as discussed we should probably move the math inside `asp.py`, since `clingo` isn't particularly good at that. For instance `trilinos` on `develop`:
```
Time:
    setup:         7.0143
    load:          0.0202
    ground:        2.5024
    solve:         3.2076
Total: 12.9201
```
while with this branch:
```
Time:
    setup:         6.9622
    load:          0.0203
    ground:        2.5175
    solve:         15.6529
Total: 25.3843
```

Reply:
I think this needs to be fixed before we can merge.